File: test_optimize.py
import time
import math

def estimate_pitch(frame, sample_rate):
    min_period = int(sample_rate / 600)
    max_period = int(sample_rate / 60)

    best_lag = 0
    best_corr = float('-inf')

    # Ensure lag doesn't exceed frame length
    max_lag = min(max_period, len(frame) - 1)

    for lag in range(min_period, max_lag + 1):
        corr = 0.0
        # for i = 0 to length - lag
        limit = len(frame) - lag
        for i in range(limit):
            corr += frame[i] * frame[i + lag]
            
        if corr > best_corr:
            best_corr = corr
            best_lag = lag

    if best_corr <= 0 or best_lag == 0:
        return 0
    return sample_rate / best_lag

def main():
    sample_rate = 16000
    frame_len = 16000
    frame = [0.0] * frame_len
    
    # Generate a sine wave at 440Hz (A4 note)
    for i in range(frame_len):
        frame[i] = math.sin(2 * math.pi * 440 * i / sample_rate)

    print("Starting Python Benchmark (estimatePitch)...")
    iterations = 50 # Run 50 times to get a good average

    start_time = time.time()
    result = 0
    for _ in range(iterations):
        result = estimate_pitch(frame, sample_rate)
    end_time = time.time()

    total_time_ms = (end_time - start_time) * 1000
    avg_time_ms = total_time_ms / iterations

    print(f"[Python] Result Pitch: {result:.2f} Hz")
    print(f"[Python] Total Time for {iterations} iterations: {total_time_ms:.2f} ms")
    print(f"[Python] Average Time per iteration: {avg_time_ms:.2f} ms")

if __name__ == "__main__":
    main()

File: test_optimize.js
const { performance } = require('perf_hooks');

function estimatePitch(frame, sampleRate) {
  const minPeriod = Math.floor(sampleRate / 600);  // 600 Hz max
  const maxPeriod = Math.floor(sampleRate / 60);   // 60 Hz min

  let bestLag    = 0;
  let bestCorr   = -Infinity;

  for (let lag = minPeriod; lag <= maxPeriod && lag < frame.length; lag++) {
    let corr = 0;
    for (let i = 0; i < frame.length - lag; i++) {
      corr += frame[i] * frame[i + lag];
    }
    if (corr > bestCorr) {
      bestCorr = corr;
      bestLag  = lag;
    }
  }

  if (bestCorr <= 0 || bestLag === 0) return 0;
  return sampleRate / bestLag;
}

// Create a dummy audio frame (1 second of audio at 16000Hz)
const sampleRate = 16000;
const frame = new Float32Array(16000);
for (let i = 0; i < frame.length; i++) {
  // Generate a sine wave at 440Hz (A4 note)
  frame[i] = Math.sin(2 * Math.PI * 440 * i / sampleRate);
}

console.log("Starting JavaScript Benchmark (estimatePitch)...");
const iterations = 50; // Run 50 times to get a good average

const start = performance.now();
let result = 0;
for (let i = 0; i < iterations; i++) {
  result = estimatePitch(frame, sampleRate);
}
const end = performance.now();

const totalTimeMs = (end - start);
const avgTimeMs = totalTimeMs / iterations;

console.log(`[JavaScript] Result Pitch: ${result.toFixed(2)} Hz`);
console.log(`[JavaScript] Total Time for ${iterations} iterations: ${totalTimeMs.toFixed(2)} ms`);
console.log(`[JavaScript] Average Time per iteration: ${avgTimeMs.toFixed(2)} ms`);
